doc/md/fundamentals/7-messaging.md

```diff
@@ -0,0 +1,170 @@
+---
+sidebar_position: 7
+---
+
+# Messaging
+
+## Messaging restrictions
+
+ICP enforces rules on when and how [canisters](https://internetcomputer.org/docs/building-apps/essentials/canisters) communicate. Motoko has static (compile-time) messaging restrictions to prevent certain execution errors.
+
+| Restriction | Reason |
+|-------------|--------|
+| A canister cannot send messages during installation. | Prevents execution errors at deployment. |
+| A query function cannot send messages. | Query functions run locally and do not trigger updates. |
+| Shared functions cannot be called in a synchronous context. | Shared calls require `async` execution. |
+| Only async contexts support [error handling](https://internetcomputer.org/docs/motoko/fundamentals/error-handling) (`try/catch`). | Messaging errors only occur in async contexts. |
+
+In Motoko, an expression is in an async context if it appears in an `async` function. Query functions are 'ready only' functions, so they do not create an async context and cannot use `await` or send messages.
+
+```motoko no-repl
+actor Counter {
+    stable var count : Nat = 0;
+
+    public shared func increment() : async () {
+        count += 1;
+    };
+
+    public query func getCount() : async Nat {
+        return count;  // Allowed: No state change
+    };
+
+    public func invalidCall() : Nat {
+        return await getCount();  // Error: Query function cannot be awaited in a sync function
+    };
+};
+```
+
+## Message inspection
+
+Canisters can inspect and filter incoming messages before execution to prevent spam or malicious requests. This is done using the `inspect` system function, which can accept or reject messages based on criteria like caller identity, message size, or function type. Messages are filtered before execution, preventing unnecessary cycles costs. Anonymous calls can be blocked to prevent spam, and the size of a request can be checked before processing.
```

Comment:
```suggestion
Canisters can inspect and filter incoming replicated messages before execution to prevent spam or malicious requests. This is done using the `inspect` system function, which can accept or reject messages based on criteria like caller identity, message size, or function type. Messages are filtered before execution, preventing unnecessary cycles costs. Anonymous calls can be blocked to prevent spam, and the size of a request can be checked before processing.
```